### PR TITLE
[etherrain] Etherrain Binding initial contribution 

### PR DESCRIPTION
**Pull request by dfad1469**
This is an initial commit of the bind for the Etherrain controller.

http://www.quicksmart.com/qs_etherrain.html

The binding exposes the execution of 7-8 zones as well as the state of the optional rain sensor. 

Signed-off-by: Joe Inkenbrandt [openhab@inkenbrandt.com] (github: dfad149)

### CODE REVIEW

File: addons/binding/org.openhab.binding.etherrain/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+/target/
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/.gitignore]: please remove this file

File: addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml [hunk 2]
```diff
@@ -0,0 +1,155 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="etherrain" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+    xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd">
+
+    <!-- The EtherRain Thing Type -->
+    <thing-type id="etherrain">
+        <label>EtherRain</label>
+        <description>This is the stand alone EtherRain device that allows sprinkler control</description>
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml]:
```suggestion
        <description>This is a stand alone EtherRain device that allows sprinkler control</description>
```

File: addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                <label>Zone 7 Run Time</label>
+                <description>The amount of time, in minutes, that EtherRain will run zone 7 in program</description>
+                <default>0</default>
+            </parameter>
+            <parameter name="zoneOnTime8" type="integer">
+                <label>Zone 8 Run Time</label>
+                <description>The amount of time, in minutes, that EtherRain will run zone 8 in program</description>
+                <default>0</default>
+            </parameter>
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                <default>0</default>
+            </parameter>
+            <parameter name="zoneOnTime8" type="integer">
+                <label>Zone 8 Run Time</label>
+                <description>The amount of time, in minutes, that EtherRain will run zone 8 in program</description>
+                <default>0</default>
+            </parameter>
+            
+        </config-description>
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                <label>Clear</label>
+                <description>Clears the currently running program</description>
+            </channel>
+        </channels>
+
+        <config-description>
+            <parameter name="hostname" type="text" required="true">
+                <label>Host Name</label>
+                <description>The host name or IP address of the EtherRain Web API interface</description>
+                <default>192.168.1 204</default>
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml]: looks like an invalid address

File: addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            </parameter>
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml]:
```suggestion
                <context>network-address</context>
            </parameter>
```

File: addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+                <description>Commands EtherRain device to begin watering program</description>
+            </channel>
+            <channel id="clear" typeId="execute">
+                <label>Clear</label>
+                <description>Clears the currently running program</description>
+            </channel>
+        </channels>
+
+        <config-description>
+            <parameter name="hostname" type="text" required="true">
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/ESH-INF/thing/thing-types.xml]: documentation says `host` (which would be in-line with other bindings)

File: addons/binding/org.openhab.binding.etherrain/README.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+The second is the initial delay and on-time for each zone when an exectute command is issued. 
+
+## Thing Configuration
+
+This is optional, it is recommended to let the binding discover and add Etherrain controllers.
+
+To manually configure a Etherrain controller you may specify it's host name or ip ("hostname"). You can also optionally specify the unit's password ("pw"), port it is communicating on ("port") or refresh rate ("refresh") 
+
+```java
+Bridge etherrain:etherrain:BackyardSprinkler [ host="192.168.1.100"]
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/README.md]: `Thing`, check if `host` is correct (XML differs)

File: addons/binding/org.openhab.binding.etherrain/about.html [hunk 9]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/about.html]: Please have a look at other contributions. We now use a `NOTICE` file instead.

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/discovery/EtherrainDiscoveryService.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            ThingUID uid = new ThingUID(EtherRainBindingConstants.ETHERRAIN_THING,
+                    rdp.getAddress().replaceAll("[^A-Za-z0-9\\-_]", ""));
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(uid)
+                    .withLabel("Etherrain " + rdp.getType() + " " + rdp.getUnqiueName())
+                    .withProperty("hostname", rdp.getAddress()).withProperty("port", rdp.getPort()).build();
+            thingDiscovered(discoveryResult);
+        } else {
+            logger.debug("Nothing responded to request");
+        }
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/discovery/EtherrainDiscoveryService.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/discovery/EtherrainDiscoveryService.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+}
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/discovery/EtherrainDiscoveryService.java]: NL at end of file. Please check the static code analysis report, there mioght be more code-style issues.

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/handler/EtherRainHandler.java [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+
+        try {
+            EtherRainStatusResponse response = device.commandStatus();
+            if (response == null) {
+                logger.debug("Command Status returned null");
+                device = null;
+                updateStatus(ThingStatus.OFFLINE);
+                return false;
+            }
+        } catch (Exception e) {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/handler/EtherRainHandler.java]: Do you really need to catch `Exception` here? It is usally better to be more specific.

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/handler/EtherRainHandler.java [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+
+        updateStatus(ThingStatus.ONLINE);
+
+        return true;
+    }
+
+    private void startUpdateJob() {
+        logger.debug("Starting Etherrain Update Job");
+        updateJob = scheduler.scheduleWithFixedDelay(() -> {
+            updateBridge();
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/handler/EtherRainHandler.java]:
```suggestion
  updateJob = scheduler.scheduleWithFixedDelay(this::updateBridge);
```

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/handler/EtherRainHandler.java [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+        logger.debug("Stopping Etherrain Update Job");
+        if (updateJob != null && !updateJob.isDone()) {
+            updateJob.cancel(false);
+        }
+
+        updateJob = null;
+    }
+
+    private boolean updateBridge() {
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/handler/EtherRainHandler.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/handler/EtherRainHandler.java [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+                logger.debug("Could not connect to Etherrain device.");
+                return false;
+            }
+        }
+
+        EtherRainStatusResponse response;
+
+        try {
+            response = device.commandStatus();
+        } catch (Exception e) {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/handler/EtherRainHandler.java]: Please do not catch `Exception` (see above)

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/EtherRainConfiguration.java [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    public static final String BINDING_ID = "etherrrain";
+
+    public static final ThingTypeUID ETHERRAIN_THING_TYPE = new ThingTypeUID(BINDING_ID, "etherrain");
+
+    public static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Collections.singleton(ETHERRAIN_THING_TYPE);
+
+    /**
+     * Hostname of the OpenSprinkler API.
+     */
+    public String hostname = null;
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/EtherRainConfiguration.java]: why don't you set the default host here? Also check that the documentation refers to `host`

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+
+    private final static int HTTP_TIMEOUT = 1000;
+
+    private final static int BROADCAST_TIMEOUT = 80;
+
+    private String address = null;
+    private int port = 80;
+    private String password = "pw";
+
+    public EtherRainCommunication() {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: do you need this constructor? it seems a bit strange that `address=null` should be any useful. if unsued, remove it. you can then remove the field initializations, too.

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+
+    private final static String ETHERRAIN_USERNAME = "admin";
+
+    private final static int HTTP_TIMEOUT = 1000;
+
+    private final static int BROADCAST_TIMEOUT = 80;
+
+    private String address = null;
+    private int port = 80;
+    private String password = "pw";
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: is `pw` a default password? if not, don't initialize with bogus content.

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+        }
+
+        Iterator<String> i = responseList.iterator();
+
+        EtherRainStatusResponse response = new EtherRainStatusResponse();
+
+        while (i.hasNext()) {
+            String line = i.next();
+
+            if (line.matches(RESPONSE_STATUS_PATTERN)) {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: maybe log if something unexcpected was received? might help debugging.

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+        this.password = password;
+    }
+
+    public static EtherRainUdpResponse autoDiscover() {
+        return updBroadcast();
+    }
+
+    private static final String RESPONSE_STATUS_PATTERN = "^\\s*(un|ma|ac|os|cs|rz|ri|rn):\\s*([a-zA-Z0-9\\.]*)(\\s*<br>)?";
+
+    public EtherRainStatusResponse commandStatus() throws Exception {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: Don't do that. Inherit a new `EtherRainExcpetion` from `Exception`. You can be more specific in what ypu catch then (see my remarks above). An example can be found here https://github.com/openhab/openhab2-addons/blob/master/bundles/org.openhab.binding.onewire/src/main/java/org/openhab/binding/onewire/internal/OwException.java.

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+                        break;
+                    case "ri":
+                        response.setLastActiveValue(Integer.parseInt(status));
+                        break;
+                    case "rn":
+                        response.setRainSensor(Integer.parseInt(status) == 1 ? true : false);
+                        break;
+                }
+            }
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+    public boolean commandIrrigate(int delay, int zone1, int zone2, int zone3, int zone4, int zone5, int zone6,
+            int zone7, int zone8) {
+
+        try {
+            if (sendGet("result.cgi?xi=" + delay + ":" + zone1 + ":" + zone2 + ":" + zone3 + ":" + zone4 + ":" + zone5
+                    + ":" + zone6 + ":" + zone7 + ":" + zone8) == null) {
+                return false;
+            }
+        } catch (Exception e) {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: see above

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 23]
```diff
@@ -0,0 +1,257 @@
...
+
+        return true;
+    }
+
+    public boolean commandClear() {
+        try {
+            if (sendGet("/result.cgi?xr") == null) {
+                return false;
+            }
+        } catch (Exception e) {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: see above

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 24]
```diff
@@ -0,0 +1,257 @@
...
+
+        return true;
+    }
+
+    public boolean commandLogin() throws Exception {
+        return sendGet("/ergetcfg.cgi?lu=" + ETHERRAIN_USERNAME + "&lp=" + password) != null;
+    }
+
+    public boolean commandLogout() {
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 25]
```diff
@@ -0,0 +1,257 @@
...
+            return false;
+        }
+
+        return true;
+    }
+
+    private final String USER_AGENT = "Mozilla/5.0";
+
+    private List<String> sendGet(String command) throws Exception {
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 26]
```diff
@@ -0,0 +1,257 @@
...
+            }
+
+        }
+
+        return response;
+    }
+
+    public boolean commandIrrigate(int delay, int zone1, int zone2, int zone3, int zone4, int zone5, int zone6,
+            int zone7, int zone8) {
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 27]
```diff
@@ -0,0 +1,257 @@
...
+        this.address = address;
+        this.port = port;
+        this.password = password;
+    }
+
+    public static EtherRainUdpResponse autoDiscover() {
+        return updBroadcast();
+    }
+
+    private static final String RESPONSE_STATUS_PATTERN = "^\\s*(un|ma|ac|os|cs|rz|ri|rn):\\s*([a-zA-Z0-9\\.]*)(\\s*<br>)?";
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: move to top of class

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 28]
```diff
@@ -0,0 +1,257 @@
...
+                return false;
+            }
+        } catch (Exception e) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private final String USER_AGENT = "Mozilla/5.0";
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: move to top of class

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 29]
```diff
@@ -0,0 +1,257 @@
...
+        BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
+        String inputLine;
+
+        while ((inputLine = in.readLine()) != null) {
+            rVal.add(inputLine);
+        }
+        in.close();
+
+        return rVal;
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 30]
```diff
@@ -0,0 +1,257 @@
...
+    }
+
+    private static final String BROADCAST_RESPONSE_DISCOVER_PATTERN = "eviro t=(\\S*) n=(\\S*) p=(\\S*) a=(\\S*)";
+
+    private static EtherRainUdpResponse updBroadcast() {
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 31]
```diff
@@ -0,0 +1,257 @@
...
+        while ((inputLine = in.readLine()) != null) {
+            rVal.add(inputLine);
+        }
+        in.close();
+
+        return rVal;
+
+    }
+
+    private static final String BROADCAST_RESPONSE_DISCOVER_PATTERN = "eviro t=(\\S*) n=(\\S*) p=(\\S*) a=(\\S*)";
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: move to top of class

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 32]
```diff
@@ -0,0 +1,257 @@
...
+        try {
+            // Open a random port to send the package
+            c = new DatagramSocket();
+            c.setSoTimeout(BROADCAST_TIMEOUT);
+            c.setBroadcast(true);
+
+            byte[] sendData = BROADCAST_DISCOVERY_MESSAGE.getBytes();
+
+            // Try the 255.255.255.255 first
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java [hunk 33]
```diff
@@ -0,0 +1,257 @@
...
+            int portBC = Integer.parseInt(message.replaceAll(BROADCAST_RESPONSE_DISCOVER_PATTERN, "$3"));
+
+            c.close();
+
+            // NOTE: Version 3.77 of API states that Additional Parameters (a=) are not used on EtherRain
+            return new EtherRainUdpResponse(deviceTypeBC, addressBC, portBC, unqiueNameBC, null);
+        } catch (IOException ex) {
+            return null;
+        }
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainCommunication.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java [hunk 34]
```diff
@@ -0,0 +1,175 @@
...
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.etherrain.internal.api;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EtherRainStatusResponse {
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java [hunk 35]
```diff
@@ -0,0 +1,175 @@
...
+        public static OperatingStatus fromString(String text) {
+            for (OperatingStatus b : OperatingStatus.values()) {
+                if (b.status.equalsIgnoreCase(text)) {
+                    return b;
+                }
+            }
+            return null;
+        }
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java [hunk 36]
```diff
@@ -0,0 +1,175 @@
...
+    private String serviceAccount;
+
+    private OperatingStatus operatingStatus = null;
+    private CommandStatus lastCommandStatus = null;
+    private CommandResult lastCommandResult = null;
+
+    private int lastActiveValue;
+    private boolean rainSensor;
+
+    public enum OperatingStatus {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java]: please move to own file.

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java [hunk 37]
```diff
@@ -0,0 +1,175 @@
...
+                if (b.status.equalsIgnoreCase(text)) {
+                    return b;
+                }
+            }
+            return null;
+        }
+
+    }
+
+    public enum CommandStatus {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java]: move to own file

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java [hunk 38]
```diff
@@ -0,0 +1,175 @@
...
+            for (CommandStatus b : CommandStatus.values()) {
+                if (b.status.equalsIgnoreCase(text)) {
+                    return b;
+                }
+            }
+            return null;
+        }
+    }
+
+    public enum CommandResult {
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java]: move to own file

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java [hunk 39]
```diff
@@ -0,0 +1,175 @@
...
+    public boolean isRainSensor() {
+        return rainSensor;
+    }
+
+    public void setRainSensor(boolean rainSensor) {
+        this.rainSensor = rainSensor;
+    }
+
+    public void printDebug() {
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java]: remove empty line

File: addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java [hunk 40]
```diff
@@ -0,0 +1,175 @@
...
+        logger.debug("Mac Address: " + macAddress);
+        logger.debug("Service Account: " + serviceAccount);
+
+        logger.debug("Operating Status: " + operatingStatus.toString());
+        logger.debug("Last Command Status: " + lastCommandStatus.toString());
+        logger.debug("Last Command Result: " + lastCommandResult.toString());
+
+        logger.debug("Last Active Valve: " + lastActiveValue);
+        logger.debug("Rain Sensor: " + rainSensor);
+
```
Comment (J-N-K) [inline addons/binding/org.openhab.binding.etherrain/src/main/java/org/openhab/binding/etherrain/internal/api/EtherRainStatusResponse.java]: remove empty line

**Comment by J-N-K** [comment]
I have another question (forgotten before): The documentation states "seven or eight" channels. the thing always has eight channels. Can you determine the channel number from some sort of "modelnumber" or something like that? We could then remove channel 8 on the devices with only seven channels.

**Comment by dfad1469** [comment]
Since I had to move this to the new build system, I did it on another branch. I am going to close the PR and open a new one on the correct branch.

**Pull request closed without merging**